### PR TITLE
DrivAerML: 4L/640d mid-width + gc=1.0 — between 512d and 768d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Width scaling is the dominant capacity lever on DrivAerML (4L/384d→4L/512d gave 8.1% improvement). guts #2890 is testing 4L/768d (ultra-wide). This experiment fills the gap: 4L/640d with 10 attention heads (640/10=64 dim/head, same as 512d/8 heads). The 640d width provides 25% more capacity than 512d while maintaining the same per-head dimension. If 768d is too wide (slower epochs, fewer total epochs), 640d may hit the optimal capacity-vs-throughput tradeoff. Combined with gc=1.0 recipe transfer.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications.**

Run 1 — DrivAerML 4L/640d + gc=1.0 (10 heads):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 640 --model-heads 10 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group chrome-640d-sweep
```

Run 2 — DrivAerML 4L/640d + gc=1.0 + WD=1e-2 (compound):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 640 --model-heads 10 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group chrome-640d-sweep
```

Run 3 — DrivAerML 4L/640d + gc=1.0 + T_max=10 (faster cycling at wider width):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 640 --model-heads 10 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group chrome-640d-sweep
```

Run 4 — AirfRANS cross-dataset (golden config replication):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --seed 45 \
  --wandb-group chrome-640d-sweep
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML and per-epoch timing (min/epoch). Also report `val_primary/surface_mse` for AirfRANS.

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, 4L/512d, ~0.67 min/epoch, 267 epochs)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, epoch 349)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chrome:

## Results

### DrivAerML — 4L/640d gc=1.0 Sweep

**Critical finding: All three DrivAerML 4L/640d runs eventually diverged catastrophically.** The 640d width (10 heads, 64 dim/head) is fundamentally unstable on DrivAerML — gradient norms spike, grad_clip saturates (350+/394 steps), and the loss/metrics blow up irreversibly.

| Run | Config | Best surface_rel_l2_pct | Best Epoch | Diverged At | Grad Clip (stable) | W&B |
|-----|--------|------------------------|------------|-------------|-------------------|-----|
| 1 | gc=1.0, no WD, T_max=30 | **8.71%** | 84 | epoch 90 | 20-30 → 354 | \`wn9e7pyq\` |
| 2 | gc=1.0, WD=1e-2, T_max=30 | **6.52%** | 125 | epoch 147 | 2-15 → 381 | \`66q5d00j\` |
| 3 | gc=1.0, no WD, T_max=10 | **6.92%** | 130 | epoch 153 | 18-53 → 387 | \`9vyw8031\` |

**Baseline: 4.619%** (PR #2691, 4L/512d, 267 epochs) · **External target: 3.71%** (AB-UPT)

**Timing:** ~0.75 min/epoch (vs ~0.67 min/epoch for 512d baseline — 12% slower)
**Peak VRAM:** ~7,374 MiB per GPU

#### Commands Used

Run 1:
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 640 --model-heads 10 --grad-clip 1.0 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group chrome-640d-sweep --agent chrome --wandb-name "chrome/4L-640d-gc1.0"
```

Run 2:
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 640 --model-heads 10 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group chrome-640d-sweep --agent chrome --wandb-name "chrome/4L-640d-gc1.0-wd1e2"
```

Run 3:
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 640 --model-heads 10 --grad-clip 1.0 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group chrome-640d-sweep --agent chrome --wandb-name "chrome/4L-640d-gc1.0-tmax10"
```

### AirfRANS — Cross-Dataset gc=1.0 + WD=1e-2 Transfer

| Run | Config | Best surface_mse | Best Epoch | Status | W&B |
|-----|--------|-----------------|------------|--------|-----|
| 4 | 2L/256d, gc=1.0, WD=1e-2, seed=45 | **0.001193** | 237 | **still running** | `xks7fw7l` |

**Baseline: 0.001236** (PR #2828, epoch 349) · **Result: 3.5% improvement, new best**

**Timing:** ~0.49 min/epoch
**Peak VRAM:** ~30,153 MiB

```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --seed 45 --wandb-group chrome-640d-sweep --agent chrome --wandb-name "chrome/airfrans-2L256d-gc1.0-wd1e2-seed45"
```

### What Happened

**DrivAerML 640d divergence is the dominant finding.** All three runs followed the same pattern:
1. Stable training with good convergence for 80-150 epochs
2. Sudden onset of gradient instability — grad_clip events spike from normal (2-50) to saturated (350+/394)
3. Grad norm explodes (0.4 → 4.6 → 8.5 → 19.5), loss triples, metric blows up
4. Irreversible within 2-3 epochs of onset

**Weight decay (WD=1e-2) delayed divergence by ~57 epochs** (Run 1: ep90 vs Run 2: ep147) and allowed a better peak metric (6.52% vs 8.71%), but could not prevent it. Faster cosine cycling (T_max=10 in Run 3) also delayed divergence slightly to ep153 and achieved 6.92%.

**The 640d width appears to be in a regime where capacity exceeds what gc=1.0 alone can stabilize.** The 512d baseline ran 267 epochs without divergence. The extra 25% capacity at 640d creates weight growth that eventually overwhelms the gradient clipping threshold.

**AirfRANS cross-dataset transfer succeeded.** The gc=1.0 + WD=1e-2 recipe on a smaller model (2L/256d) beat the AirfRANS baseline by 3.5%, reaching 0.001193 surface_mse at epoch 237. This run is still training and may improve further.

### Suggested Follow-ups

1. **Stronger regularization for 640d**: Try WD=5e-2 or WD=1e-1 — 1e-2 delayed but didn't prevent divergence. The 640d width may need substantially more regularization than 512d.
2. **Lower learning rate for 640d**: Try lr=2e-4 or lr=1e-4 — the instability may be a learning rate issue that compounds with increased model capacity.
3. **Gradient clipping value**: gc=0.5 or gc=0.3 might prevent the catastrophic phase — when gc=1.0 clips nearly every step, the effective training dynamic changes.
4. **Warmup**: A linear warmup phase (e.g. 10-20 epochs) before full LR could prevent the initial weight growth that seeds later divergence.
5. **576d intermediate width**: If 512d is stable at 267 epochs and 640d diverges at 90-150, there may be a stability boundary in between. 576d (9 heads × 64 dim/head) would probe this.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2917 because ADVISOR: Closing — DrivAerML 4L/640d is unstable at lr=5e-4+gc=1.0:

| Run | Config | Best | Epoch | Fate |
|---|---|---|---|---|
| Run 1 | 640d+gc=1.0 | 8.71% | ep84 | crashed ep90 |
| Run 2 | 640d+gc=1.0+WD=1e-2 | 6.52% | ep125 | crashed ep147 |
| Run 3 | 640d+gc=1.0+T_max=10 | 6.92% | ep130 | crashed ep153 |

All three diverged catastrophically. The 512d→640d transition crosses a stability boundary at current hyperparameters. WD=1e-2 bought ~57 more stable epochs but couldn't prevent eventual explosion.

**Notable AirfRANS result:** Your Run 4 (2L/256d golden AF config at seed=45) hit 0.001193 at ep237 — 3.5% below the 0.001236 baseline. This is interesting but the run is oscillating badly in late epochs (0.004-0.009), suggesting T_max=5 instability. Worth monitoring but not stable enough to merge from this PR.

Your follow-up suggestions were sharp — lower LR (2e-4) for 640d is the right stabilization lever, and 576d (9×64) is a more conservative width step. I'll assign you a fresh direction. Thank you for the thorough analysis..
